src/node/state.rs

```diff
@@ -0,0 +1,182 @@
+    },
+    InsufficientBalance {
+        account: AccountState,
+        amount: Amount,
+    },
+    BalanceOverflow {
+        account: AccountState,
+        amount: Amount,
+    },
+    NonceOverflow,
+}
+
+#[derive(Clone, Debug)]
+pub struct AccountState {
+    balance: Amount,
+    next_nonce: Nonce,
+}
+
+impl AccountState {
+    pub fn new(balance: Amount) -> Self {
+        AccountState {
+            balance,
+            next_nonce: 0,
+        }
+    }
+
+    pub fn balance(&self) -> &Amount {
+        &self.balance
+    }
+
+    pub fn next_nonce(&self) -> &Nonce {
+        &self.next_nonce
+    }
+
+    pub fn transfer(
+        &mut self,
+        to_account: &mut Self,
+        amount: Amount,
+    ) -> Result<(), TransactionExecutionError> {
+        if amount > self.balance {
+            return Err(TransactionExecutionError::InsufficientBalance {
+                account: self.clone(),
+                amount,
+            });
+        }
+
+        if amount > Amount::MAX - self.balance {
+            return Err(TransactionExecutionError::BalanceOverflow {
+                account: to_account.clone(),
+                amount,
+            });
+        }
+
+        if self.next_nonce == Nonce::MAX {
+            return Err(TransactionExecutionError::NonceOverflow);
+        }
+
+        self.balance -= amount;
+        self.next_nonce += 1;
+        to_account.balance += amount;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct RollupState {
+    accounts: HashMap<Address, AccountState>,
+    batch_count: usize,
+}
+
+impl RollupState {
+    pub fn new() -> Self {
+        Self {
+            accounts: HashMap::new(),
+            batch_count: 0,
+        }
+    }
+
+    pub fn accounts(&self) -> &HashMap<Address, AccountState> {
+        &self.accounts
+    }
+
+    pub fn batch_count(&self) -> &usize {
+        &self.batch_count
+    }
+
+    pub fn apply_transaction(
+        &mut self,
+        transaction: &Transaction,
+    ) -> Result<(), TransactionExecutionError> {
+        match transaction {
+            Transaction::Enter(Enter { account, amount }) => {
+                self.accounts
+                    .entry(*account)
+                    .or_insert(AccountState::new(0))
+                    .balance += amount;
+            }
+            Transaction::Transfer(
+                transfer @ Transfer {
+                    from: from_address,
+                    to: to_address,
+                    amount,
+                    nonce: _,
+                },
+            ) => {
+                self.validate_addresses(transfer)?;
+
+                let mut from_account = self.accounts.remove(from_address).unwrap();
+                let mut to_account = self.accounts.remove(to_address).unwrap();
+                from_account.transfer(&mut to_account, *amount)?;
+                self.accounts.insert(*from_address, from_account);
+                self.accounts.insert(*to_address, to_account);
```

Comment:
Getting access to two mapping elements at the same time is surprisingly annoying in Rust if you need to modify both.

There's [`HashMap::get_many_mut()`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get_many_mut), which would make it a little more concise:
```rust
let Some([from_account, fo_account]) = self.accounts.get_many_mut([from_address, to_address]);
from_account.transfer(to_account, *amount);
```
It's only available in nightly releases of the compiler though.